core/src/test/java/org/apache/calcite/sql2rel/RelFieldTrimmerTest.java

```diff
@@ -183,4 +187,44 @@
     assertThat(trimmed, hasTree(expected));
   }
 
+  /** Test case for
+   * <a href="https://issues.apache.org/jira/browse/CALCITE-4055">[CALCITE-4055]
+   * RelFieldTrimmer loses hints</a>. */
+  @Test void testJoinWithHints() {
+    final RelHint noHashJoinHint = RelHint.builder("NO_HASH_JOIN").build();
+    final RelBuilder builder = RelBuilder.create(config().build());
```

Comment:
Great fix, can we also add a test case for the aggregate ? thanks for that ~

Reply:
We can also promote the `RelBuilder#hints` like i said in the JIRA.

Reply:
Thanks, I will add more tests.
What do you mean by "promoting" `RelBuilder#hints` ?

Reply:
Disregard my question, I already saw what you mean "if the passed in hints is empty, just returns the builder early". Ok, I will do that.

Reply:
changes committed, please check.